src/bookshelf/src/bksw_sql.erl

```diff
@@ -0,0 +1,274 @@
+%% @copyright 2015 Chef, Inc. All Rights Reserved
+%% @author Mark Anderson <mark@chef.io>
+%%
+%% Licensed to the Apache Software Foundation (ASF) under one or more
+%% contributor license agreements.  See the NOTICE file distributed
+%% with this work for additional information regarding copyright
+%% ownership.  The ASF licenses this file to you under the Apache
+%% License, Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain a copy of
+%% the License at http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+%% implied.  See the License for the specific language governing
+%% permissions and limitations under the License.
+-module(bksw_sql).
+
+-include("internal.hrl").
+
+%-ifdef(namespaced_types). %% workaround for sqerl complaint
+-type dict() :: dict:dict().
+%-endif.
+
+-export([ statements/1,
+          ping/0,
+          create_bucket/1,
+          delete_bucket/1,
+          find_bucket/1,
+          list_buckets/0,
+          list_bucket/1,
+          find_file/1,
+          find_file/2,
+          create_file/2,
+          create_file_with_data/3,
+          update_file_with_data/2,
+          delete_file/2, %% deprecate
+          delete_file/1,
+          rename_file_with_overwrite/3,
+          insert_file_data/0,
+          delete_file_data/1,
+          add_file_chunk/3,
+          mark_file_done/6,
+          get_chunk_data/2,
+          replace_chunk_data/1,
+          init_upload_state/0,
+          update_upload_state/2,
+          finalize_upload_state/2
+        ]).
+
+-include_lib("sqerl/include/sqerl.hrl").
+
+ping() ->
+    case sqerl:select(ping, [], first_as_scalar, [ping]) of
+        {ok, <<"pong">>} -> ok;
+        {error, Reason} -> {error, Reason}
+    end.
+
+%% @doc bookshelf queries.
+%% Referenced by sys.config in {sqerl, [{ prepared_staements...
+statements(_) ->
+    Path = filename:join([code:priv_dir(bookshelf), "pgsql_statements.config"]),
+    {ok, Statements} = file:consult(Path),
+    Statements.
+
+create_bucket(BucketName) ->
+    case sqerl:statement(insert_bucket, [BucketName], count) of
+        {ok, 1} ->
+            ok;
+        {conflict, _} ->
+            error("Never should happen");
+        Error ->
+            Error
+    end.
+
+delete_bucket(BucketName) ->
+    case sqerl:statement(delete_bucket, [BucketName], count) of
+        {ok, 1} ->
+            ok;
+        Error ->
+            Error
```

Comment:
There is some inconsistency around whether functions in the module return `{error, Error}` or `Error`.  Similarly, for `{ok Result}` vs `Result`.  My preference would be that we *always* return tagged tuples so that callers can use a pattern match to ensure errors don't unexpectedly propagate unhandled too far from their source.